### ISSUE TITLE
Bug: Query on inherited vertex property index seems to fail in OrientDB 2.0.6

### ISSUE THREAD

**Issue by vvoid** [orientechnologies/orientdb#4065]
Please review the attached test case and the comments below.

```java
package tests;

import com.orientechnologies.orient.core.metadata.schema.OClass;
import com.orientechnologies.orient.core.metadata.schema.OType;
import com.orientechnologies.orient.core.sql.OCommandSQL;
import com.tinkerpop.blueprints.Vertex;
import com.tinkerpop.blueprints.impls.orient.OrientGraph;
import com.tinkerpop.blueprints.impls.orient.OrientGraphFactory;
import com.tinkerpop.blueprints.impls.orient.OrientGraphNoTx;
import com.tinkerpop.blueprints.impls.orient.OrientVertex;
import com.tinkerpop.blueprints.impls.orient.OrientVertexType;
import org.testng.annotations.AfterClass;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

/**
 * Demonstrates a problem with an inherited vertex property index.
 *
 * 
The problem is, that a vertex in the database is not found
by a query on an index property which should actually 
find the vertex. If a plocal database is used instead of the 
in memory database the problem persists and can be
investigated with the studio. There, the vertex 
can also not be found using a query using the indexed 
property.

 *
 * Please note, the only difference between fail1() and ok2()
is the presence of the index. If the index is removed as in ok2()
the problem disappears. Also, other index types (FULLLTEXT)
do not have the same problem.
 *
 * A very similar setup in ok1() does also not show this problem.
 *
 * Please investigate...
 *
 * @author root
 */
public class IndexProblemDemo1 {

  public IndexProblemDemo1() {
  }

  public static void setupSchemaFails(OrientGraphFactory factory) {
    OrientGraphNoTx graphNoTx = factory.getNoTx();

    OrientVertexType vertexType;
    {
      vertexType = graphNoTx.createVertexType("s1", "V");
      vertexType.setAbstract(true);
      vertexType.createProperty("name", OType.STRING)
        .setMandatory(false)
        .setNotNull(false)
        .createIndex(OClass.INDEX_TYPE.NOTUNIQUE_HASH_INDEX) // this line breaks it all ?!?
        // btw other index types have the same problem
        ;

      {
        vertexType = graphNoTx.createVertexType("s2", "s1");
        vertexType.getProperty("name")
          .setMandatory(true)
          .setNotNull(true);
      }

    }

    graphNoTx.shutdown();

  }

  public static void setupSchemaOk(OrientGraphFactory factory) {
    OrientGraphNoTx graphNoTx = factory.getNoTx();

    OrientVertexType vertexType;
    {
      vertexType = graphNoTx.createVertexType("s1", "V");
      vertexType.setAbstract(true);
      vertexType.createProperty("name", OType.STRING)
        .setMandatory(false)
        .setNotNull(false) 
        //.createIndex(OClass.INDEX_TYPE.NOTUNIQUE_HASH_INDEX) // removal fixes the query
        //.createIndex(OClass.INDEX_TYPE.FULLTEXT) // this index type seems to be unaffected
        ;

      {
        vertexType = graphNoTx.createVertexType("s2", "s1");
        vertexType.getProperty("name")
          .setMandatory(true)
          .setNotNull(true);
      }

    }

    graphNoTx.shutdown();

  }

  public static OrientVertex createVertex(OrientGraphFactory factory, String className, Object... fields) {
    OrientVertex vertex = null;

    OrientGraph graph = factory.getTx();

    vertex = graph.addVertex("class:" + className);
    vertex.setProperties(fields);

    graph.shutdown();

    return vertex;
  }

  @Test
  public void fail1() throws Exception {
    OrientGraphFactory factory = new OrientGraphFactory("memory:test1", "admin", "admin");

    setupSchemaFails(factory);
    createVertex(factory, "s2", "name", "notFound");

    OrientGraphNoTx graphNoTx = factory.getNoTx();

    OCommandSQL cmd = new OCommandSQL(
      " SELECT * FROM s2 "
      + " WHERE name='notFound'"
    );

    for (Vertex v : graphNoTx.getVerticesOfClass("s2")) {
      System.out.println("in database: " + ((OrientVertex) v).getRecord().toJSON());
    }

    int i = 0;
    for (OrientVertex v : (Iterable<OrientVertex>) graphNoTx.command(cmd).execute()) {
      i++;
      System.out.println(v);
    }

    if (i != 1) {
      throw new Exception("vertex not found!");
    }

  }

  @Test
  public void ok2() throws Exception {
    OrientGraphFactory factory = new OrientGraphFactory("memory:test3", "admin", "admin");

    setupSchemaOk(factory);
    createVertex(factory, "s2", "name", "notFound");

    OrientGraphNoTx graphNoTx = factory.getNoTx();

    OCommandSQL cmd = new OCommandSQL(
      " SELECT * FROM s2 "
      + " WHERE name='notFound'"
    );

    for (Vertex v : graphNoTx.getVerticesOfClass("s2")) {
      System.out.println("in database: " + ((OrientVertex) v).getRecord().toJSON());
    }

    int i = 0;
    for (OrientVertex v : (Iterable<OrientVertex>) graphNoTx.command(cmd).execute()) {
      i++;
      System.out.println(v);
    }

    if (i != 1) {
      throw new Exception("vertex not found!");
    }

  }

  @Test
  public void ok1() throws Exception {

    OrientGraphFactory factory = new OrientGraphFactory("memory:test2", "admin", "admin");

    // setup
    OrientGraphNoTx graphNoTx = factory.getNoTx();

    OrientVertexType vertexType;
    {
      vertexType = graphNoTx.createVertexType("ok1", "V");
      vertexType.setAbstract(true);
      vertexType.createProperty("name", OType.STRING)
        .setMandatory(false)
        .setNotNull(false)
        .createIndex(OClass.INDEX_TYPE.NOTUNIQUE_HASH_INDEX);

      {
        vertexType = graphNoTx.createVertexType("ok2", "ok1");
        vertexType.getProperty("name")
          .setMandatory(true)
          .setNotNull(true);
      }

    }

    graphNoTx.shutdown();

    // add vertex
    OrientGraph graph = factory.getTx();
    graph.addVertex("class:" + "ok2", "name", "found");
    graph.shutdown();

    graphNoTx = factory.getNoTx();

    OCommandSQL cmd = new OCommandSQL(
      " SELECT * FROM ok2 "
      + " WHERE name='found'"
    );

    for (Vertex v : graphNoTx.getVerticesOfClass("ok2")) {
      System.out.println("in database: " + ((OrientVertex) v).getRecord().toJSON());
    }

    int i = 0;
    for (OrientVertex v : (Iterable<OrientVertex>) graphNoTx.command(cmd).execute()) {
      i++;
      System.out.println(v);
    }

    if (i != 1) {
      throw new Exception("vertex not found!");
    }

  }

  @BeforeClass
  public static void setUpClass() throws Exception {
  }

  @AfterClass
  public static void tearDownClass() throws Exception {
  }

  @BeforeMethod
  public void setUpMethod() throws Exception {
  }

  @AfterMethod
  public void tearDownMethod() throws Exception {
  }

}
```

**Comment by hg-jt** [comment]
+1
I am seeing this same issue in 2.1.2 when using a unique index on a property that is inherited from an abstract class. If I query the index immediately after creating the record using the Java Graph API (e.g. `graph.getVertices("Foo.foo_id", id)`, it succeeds (possibly because the Vertex is cached?), but future queries will fail and on closer inspection, nothing is inserted into the index.

**Comment by luigidellaquila** [comment]
This issue is already fixed in last 2.2 GA. I'm closing this ticket.

Please reopen if you see any related problems

Thanks

Luigi
